src/OSPSuite.Infrastructure.Import/Core/DataSource.cs

```diff
@@ -153,36 +152,80 @@ public bool ValidateDataSource(IReadOnlyList<ColumnInfo> columnInfos, IDimension
-                         || errorDimension == null 
-                         || errorDimension.Name == Constants.Dimension.FRACTION)
+                     var errorDimension = errorColumn.Key.Column.Dimension;
+                     var measurementDimension = measurementColumn.Key.Column.Dimension;
+
+                     if (errorDimension == null)
+                     {
+                        for (var i = 0; i < measurementColumn.Value.Count(); i++)
+                        {
+                           if (double.IsNaN(errorColumn.Value.ElementAt(i).Measurement))
+                              continue;
+
+                           if (column.SupportedDimensions.FirstOrDefault(x => x.HasUnit(measurementColumn.Value.ElementAt(i).Unit)) !=
+                               column.SupportedDimensions.FirstOrDefault(x => x.HasUnit(errorColumn.Value.ElementAt(i).Unit)))
+                              throw new ErrorUnitException();
+                        }
+                     }
+                     else
+                     {
+                        if (errorDimension == Constants.Dimension.NO_DIMENSION
+                            || errorDimension.Name == Constants.Dimension.FRACTION)
+                           continue;
+
+                        if (measurementDimension != errorDimension)
+                           throw new ErrorUnitException();
+                     }
+                  }
+               }
+            }
+         }
+      }
+      private void validateUnitsSupportedAndSameDimension(IReadOnlyList<ColumnInfo> columnInfos)
+      {
+         foreach (var columnInfo in columnInfos)
+         {
+            foreach (var dataSet in DataSets)
+            {
+               foreach (var set in dataSet.Data)
+               {
+                  var column = set.Data.FirstOrDefault(x => x.Key.ColumnInfo.Name == columnInfo.Name);
+
+                  if (column.Key == null)
                         continue;
 
-                     for (var i = 0; i < measurementColumn.Value.Count(); i++)
+                  //if unit comes from a column
+                  if (column.Key.Column.Dimension == null)
+                  {
+                     var dimensionOfFirstUnit = columnInfo.SupportedDimensions.FirstOrDefault(x => x.HasUnit(column.Value.ElementAt(0).Unit));
+
+                     for (var i = 0; i < column.Value.Count(); i++)
                      {
-                        if (double.IsNaN(errorColumn.Value.ElementAt(i).Measurement))
+                        if (double.IsNaN(column.Value.ElementAt(i).Measurement))
                            continue;
 
-                        if (dimensionFactory.DimensionForUnit(measurementColumn.Value.ElementAt(i).Unit) !=
-                            dimensionFactory.DimensionForUnit(errorColumn.Value.ElementAt(i).Unit))
-                           return false;
+                        if (!columnInfo.SupportedDimensions.Any(x => x.HasUnit(column.Value.ElementAt(i).Unit)))
```

Comment:
I would love to see some comments as to WHAT We are checking. Same goes for the method above. These are 3 imbricatd forEach etc... very complicated to comprehend for an outsider

Reply:
comments added